### ISSUE TITLE
quiet option does not disable new worker forked message when -W -db  flags are used / in official image

### ISSUE THREAD

**Issue by markslater** [haproxy/haproxy#1320]
### Detailed Description of the Problem

When HAProxy is run with the `quiet` flag in the configuration, with and the `-W -db` flags on the command line, on start up it still outputs:
```
[NOTICE]   (1) : New worker #1 (7) forked
```
These flags are the default flags when using the [official Docker image](https://hub.docker.com/_/haproxy).

### Expected Behavior

The quiet flag should disable all `NOTICE` level output

### Steps to Reproduce the Behavior

1. Write the following configuration to e.g. `/tmp/haproxy.cfg`:
    ```
    global
      quiet
    
    defaults
    
    frontend the_frontend
    
    backend the_backend
    ```
1. Either:
    1. execute `haproxy -W -db -f /tmp/haproxy.cfg`; or
    2. execute `docker run --rm -it -v /tmp:/usr/local/etc/haproxy:ro haproxy:2.4.2-alpine`
1. observe the output

### Do you have any idea what may have caused this?

I suspect this output is being produced before the flag has been read from the configuration.

### What is your configuration?

```haproxy
global
  quiet

defaults

frontend the_frontend

backend the_backend
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.2-553dee3 2021/07/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.2.html
Running on: Linux 5.4.0-77-generic #86-Ubuntu SMP Thu Jun 17 02:35:03 UTC 2021 x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.1.1k  25 Mar 2021
Running on OpenSSL version : OpenSSL 1.1.1k  25 Mar 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.3.1 20210424

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

### Additional Information

I'm using the official Docker image, tag `2.4.2-alpine`

**Comment by wtarreau** [comment]
I agree, this looks quite odd. I could accept that a message would be emitted before parsing the `quiet` keyword but here it's about a worker being forked, this should not happen.

I see where this comes from, it was introduced very recently as a way to collect messages over the CLI, but this feels wrong in this specific case:
0a1cdcceb ("MINOR: log: do not discard stderr when starting is over"). 

I think we ought to make sure not to emit to the stderr/logs in this case but we're confronted to a stupid chicken-and-egg issue:
  - the ha_notice() function doesn't know where this is going to be sent
  - the print_message() function that decides where to send has no idea about the requested level

@a-denoyelle we definitely need to revisit this and to figure a better way to proceed. I think that print_message() should take an argument indicating the level and should do the filtering itself. But that's just an idea.

**Comment by a-denoyelle** [comment]
Indeed my patch is broken in regards to this behavior. I think it can be solved by moving the check for MODE_QUIET/VERBOSE inside print_message directly. However I'm not sure it needed to know the message level : quiet mode filters every levels without distinction, as I see it in the code and in the documentation.

**Comment by wtarreau** [comment]
I don't think your patch it technically broken, it's just that we're hijacking an error path to return progress indications and statuses to the CLI users and that it doesn't cope well with the contradicting rules we have between "quiet" for the logs/startup and the CLI responses.

I didn't remember that quiet would filter all messages, so that could indeed be easier. This would mean that the test for QUIET would be made in a different path from the usermsgs_put(). That part isn't very clear to be by the way, as I'm seeing unconditional fprintf() after that one and this is definitely an issue. I guess you need to refactor in the same rules that apply to the ha_* functions that call it, in order to decide whether or not to emit messages on stderr, and that in addition, when !STARTING we need to duplicate them to usermsgs. That's just a guess though.

**Comment by a-denoyelle** [comment]
In fact the problem is not caused by the aforementioned commit. The problem is located on haproxy.c on the line 2311

```
if (arg_mode & (MODE_DEBUG | MODE_FOREGROUND)) {
       /* command line debug mode inhibits configuration mode */
       global.mode &= ~(MODE_DAEMON | MODE_QUIET);
       global.mode |= (arg_mode & (MODE_DEBUG | MODE_FOREGROUND));
}
```

Argument '-db' enables foreground mode, which triggers the if-block location. However, I think there is maybe an error on the condition in regards with the comment. Currently, quiet mode is disabled if DEBUG or FORGROUND is activated. For me it would be more coherent to test for both mode to be activated to disable MODE_QUIET.

**Comment by wtarreau** [comment]
Indeed, it's a detail I didn't remember. The only thing is that it has *always* worked like this since `-db` was introduced 15 years ago, so I'm a bit wary of changing such a behavior right now, especially once `-db` becomes more common in master-worker mode :-/

Similarly, `quiet` mode has always hidden messages on stderr since its introduction in 2005, while not eliminating them from the logs.

We could discuss changing these, but they are definitely breaking changes for a lot of users, starting with the reporter here who would like to see important messages but not non-important ones.

Probably that the best solution would be to support an output message level (debug/diag/notice/warning/alert/none) below which output is filtered. The debug mode (`-d`) would just be one level. `-db` would default to level notice. Quiet would eliminate all messages. And one option (maybe an optional letter or digit after `-V`) would allow to adjust the verbosity level. It seems to me there exists a level between debug and diag that is "verbose" which corresponds to current `-V`.

**Comment by wtarreau** [comment]
Should we turn this to a feature request ? It's been 9 months or so without discussion.
